Improve the performance of Clifford.from_circuit()

itoko opened the pull request:
### Summary
Improve the performance of Clifford.from_circuit() by avoiding unnecessary `to_instruction` calls.

### Details and comments
As a by-product, the `clifford_circuits._append_circuit` function is refactored into two functions `clifford_circuits._append_circuit` and `clifford_circuits._append_operation` for readability.

itoko commented:
Performance improvement by this PR:

[1q-Clifford circuits test case]
With this PR: 432 ms ± 23 ms per loop (mean ± std. dev. of 20 runs, 1 loop each)
main branch: 1.49 s ± 48.6 ms per loop (mean ± std. dev. of 20 runs, 1 loop each)
[2q-Clifford circuits test case]
With this PR: 267 ms ± 15.5 ms per loop (mean ± std. dev. of 20 runs, 1 loop each)
main branch: 626 ms ± 32.8 ms per loop (mean ± std. dev. of 20 runs, 1 loop each)

Test code:
```
# common
from qiskit.circuit import QuantumCircuit
from qiskit.quantum_info import Clifford
from qiskit.quantum_info.random import random_clifford
import numpy as np
rng = np.random.default_rng(seed=123)
```
```
# 1q-Clifford circuits test case
circuits = [random_clifford(1, rng).to_circuit() for _ in range(10000)]
%timeit -r 20 [Clifford.from_circuit(circ) for circ in circuits]
```
```
# 2q-Clifford circuits test case
circuits = [random_clifford(2, rng).to_circuit() for _ in range(2000)]
%timeit -r 20 [Clifford.from_circuit(circ) for circ in circuits]
```

jakelishman commented on qiskit/quantum_info/operators/symplectic/clifford_circuits.py:
```diff
@@ -91,21 +125,13 @@ def _append_circuit(clifford, circuit, qargs=None):
     # are a single qubit Clifford gate rather than raise an exception.
     if gate.definition is None:
         raise QiskitError(f"Cannot apply Instruction: {gate.name}")
-    if not isinstance(gate.definition, QuantumCircuit):
+    if isinstance(gate.definition, QuantumCircuit):
```
This is guaranteed by the type `QuantumCircuit` - if some object is returning something else from `definition` it's the object at fault, not this function.  You should be fine to skip the error check here.

itoko replied:
I've checked the type of `gate.definition` expecting it should be `Optional[QuantumCircuit]`, and I'm surprised it seems undefined, at lease, I couldn't find any document describing it.
Adding the type definition to `gate.definition` may be beyond the scope of this PR, I'm leaning to leave those lines as they are and leave a comment for now. What do you think?

jakelishman replied:
We might not have any specific documentation, but it should be guaranteed throughout the library - there are various parts of the transpiler that look at it, and it's only allowed to be `None` or `QuantumCircuit`.  Do you know of anywhere that violates that?

itoko replied:
OK. Yeah, I don't know anywhere it is violated. I'll remove the check to be practical.

itoko replied:
Done in [e0ed50c](https://github.com/Qiskit/qiskit-terra/pull/8585/commits/e0ed50c23794a008bcc43fbfa9823df3cc8ac025)

jakelishman commented on qiskit/quantum_info/operators/symplectic/clifford_circuits.py:
```diff
@@ -39,10 +39,44 @@ def _append_circuit(clifford, circuit, qargs=None):
+
+    gate = operation
 
     # Basis Clifford Gates
     basis_1q = {
```
If you need more speed, you might want to refactor the `basis_1q` and `basis_2q` lookups into a global hashmap so they only need to be constructed once.  It probably won't affect much, but it might be a shade faster.

itoko replied:
Done in [226892c](https://github.com/Qiskit/qiskit-terra/pull/8585/commits/226892c2bf30e3362a247b593359209dc2b6c2fc)

jakelishman reviewed (commented):
This seems like a very logical refactoring, and the fact that it comes with a speedboost from avoiding copying the input is an added bonus.

jlapeyre commented on qiskit/quantum_info/operators/symplectic/clifford_circuits.py:
```diff
@@ -39,10 +39,44 @@ def _append_circuit(clifford, circuit, qargs=None):
+            new_qubits = [qargs[qubit_indices[tup]] for tup in instruction.qubits]
+            _append_operation(clifford, instruction.operation, new_qubits)
+        return clifford
+
+    raise QiskitError("The circuit must be QuantumCircuit or Instruction")
```
+1. Validation, if needed, should have happened at an earlier point. Usually `assert` is preferred if you want to catch a programming error, say for debugging, rather than a user error, that is data validation. I don't know where to find a succinct, authoritative explanation, but there are many sources online, like [this one](https://realpython.com/python-assert-statement/#when-not-to-use-assertions) that tell part of the story.

itoko replied:
Fixed in [349390d](https://github.com/Qiskit/qiskit-terra/pull/8585/commits/349390d490e1691746910ee669bdbba5969a9580)

itoko commented on qiskit/quantum_info/operators/symplectic/clifford_circuits.py:
```diff
@@ -39,10 +39,44 @@ def _append_circuit(clifford, circuit, qargs=None):
         qargs = list(range(clifford.num_qubits))
 
+    if isinstance(circuit, Instruction):
+        return _append_operation(clifford, circuit, qargs)
```
Done in [349390d](https://github.com/Qiskit/qiskit-terra/pull/8585/commits/349390d490e1691746910ee669bdbba5969a9580)

itoko commented on qiskit/quantum_info/operators/symplectic/clifford_circuits.py:
```diff
@@ -39,10 +39,44 @@ def _append_circuit(clifford, circuit, qargs=None):
+                )
+            # Get the integer position of the flat register
+            new_qubits = [qargs[qubit_indices[tup]] for tup in instruction.qubits]
+            _append_operation(clifford, instruction.operation, new_qubits)
```
Done in [349390d](https://github.com/Qiskit/qiskit-terra/pull/8585/commits/349390d490e1691746910ee669bdbba5969a9580)

jakelishman reviewed (dismissed):
(This is a fake review hold, just as a reminder for us that @alexanderivrii wants to review before merge.)

ikkoham reviewed (commented):
LGTM. I want to wait @alexanderivrii's comments.

alexanderivrii commented:
Thanks! This looks good to me. My original question to @ShellyGarion (which somehow got me assigned to this issue) is what happens when an `Instruction` without `definition` is appended to a `Clifford`, and now I see that it throws a `QiskitError` -- which totally makes sense (as it should be the responsibility of the caller of `_append_operation` to make sure that the operation can be added). 

I have just one tiny comment. In `_append_circuit` you are checking that no instruction in the circuit can have classical bits:
```
for instruction in circuit:
        if instruction.clbits:
            raise QiskitError(
                f"Cannot apply Instruction with classical bits: {instruction.operation.name}"
            )
```
Would it also make sense to check that no instruction is a conditional gate, as for instance for the following circuit
```
qc = QuantumCircuit(2, 1)
qc.h(0).c_if(0, 0)
qc.cx(0, 1)
```

itoko commented:
I think this is now ready to be merged. I rebased to the main branch to merge #7269 and added a check for conditional gate following reviewers suggestions.
Could you take a final look, @alexanderivrii @jakelishman ?

jakelishman reviewed (approved):
This still seems fine to me, and appears to address Sasha's comments as well.  Thanks for the changes - since I'm holding 0.22 up with control-flow stuff at the moment anyway, we can slip this in as the PR was in well before the deadline.
